server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -471,94 +552,162 @@ private CachePopulator getCachePopulator(String segmentId, Interval segmentInter
+      } else {
+        serverResults = getAndCacheServerResults(serverRunner, segmentsOfServerSpec);
+      }
+      return serverResults;
+    }
 
-              @Override
-              public void cleanup(Iterator<Object> iterFromMake)
-              {
-              }
-            }
+    private ServerMaybeSegmentMaybeCache<T> pickServer(SerializablePair<ServerToSegment, Optional<T>> tuple)
+    {
+      final Optional<T> maybeResult = tuple.getRhs();
+      if (maybeResult.isPresent()) {
+        return new ServerMaybeSegmentMaybeCache<T>(ALREADY_CACHED_SERVER, Optional.empty(), maybeResult);
+      }
+      final ServerToSegment serverToSegment = tuple.getLhs();
+      final QueryableDruidServer queryableDruidServer = serverToSegment.getServer().pick();
+      if (queryableDruidServer == null) {
+        log.makeAlert(
+            "No servers found for SegmentDescriptor[%s] for DataSource[%s]?! How can this be?!",
+            serverToSegment.getSegmentDescriptor(),
+            query.getDataSource()
+        ).emit();
+        return new ServerMaybeSegmentMaybeCache<T>(
+            ALREADY_CACHED_SERVER,
+            Optional.empty(),
+            Optional.empty()
         );
-        listOfSequences.add(Sequences.map(cachedSequence, pullFromCacheFunction));
       }
+      final DruidServer server = queryableDruidServer.getServer();
+      return new ServerMaybeSegmentMaybeCache<T>(
+          server,
+          Optional.ofNullable(serverToSegment.getSegmentDescriptor()),
+          Optional.empty()
+      );
     }
 
-    private void addSequencesFromServer(
-        final List<Sequence<T>> listOfSequences,
-        final SortedMap<DruidServer, List<SegmentDescriptor>> segmentsByServer
-    )
+    // This materializes the input stream in order to group it by server
+    // This method takes in the stream of cache resolved items and will group all the items by server.
+    // Each entry in the output stream contains a list whose entries' getServer is the same
+    // Each entry will either have a present segemnt descriptor or a present result, but not both
+    // Downstream consumers should check each and handle appropriately.
+    private Stream<List<ServerMaybeSegmentMaybeCache<T>>> groupCachedResultsByServer(Stream<SerializablePair<ServerToSegment, Optional<T>>> cacheResolvedStream)
     {
-      segmentsByServer.forEach((server, segmentsOfServer) -> {
-        final QueryRunner serverRunner = serverView.getQueryRunner(server);
+      return cacheResolvedStream.map(
+          this::pickServer
+      ).collect(
+          Collectors.groupingBy(ServerMaybeSegmentMaybeCache::getServer)
+      ).values(
+          // Even though we have matrialized the map, and are iterating across the values,
+          // the java.util.HashMap.ValueSpliterator does not currently support things like sizing or subsetting,
+          // so we want to convert the collection into a form conducive with embarassing parallelization down the line
```

Comment:
I don't see how the code that directly follows this comment "converts the collection into a form conducive with embarassing parallelization down the line".

This kind of comment is IMO a sign that Stream is a wrong abstraction in this place.

Could we replace some streams with ordinary collections in this class?

Reply:
My understanding of the use of Spliterators is to handle just this kind of scenario, where you can split / apply / combine work. As such, this PR includes ways to use such mechanisms in a standard java way whenever possible.

Reply:
this code got refactored into `deserializeFromCache` for the materialization components it looks like, but the comment didn't move.

The comment is still valid though, we have a fixed size map resulting out, but the standard java stream implementation doesn't yet have a "fixed map value spliterator" optimization. There is no reason to believe it wouldn't in the future though. Or that a "fixed result size grouping collector" would be unfeasible. Such optimizations or framework improvements are beyond the scope of this PR. It is better to just document the current behavior and the points where that behavior influences parts of the code base.